dev/TEAL_opcodes.md

```diff
@@ -0,0 +1,440 @@
+
+## arg
+
+- Opcode: 0x2c {uint8 arg index N}
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[N] value to stack by index
+
+## arg_0
+
+- Opcode: 0x2d 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[0] to stack
+
+## arg_1
+
+- Opcode: 0x2e 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[1] to stack
+
+## arg_2
+
+- Opcode: 0x2f 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[2] to stack
+
+## arg_3
+
+- Opcode: 0x30 
+- Pops: _None_
+- Pushes: []byte
+- push LogicSig.Args[3] to stack
+
+## txn
+
+- Opcode: 0x31 {uint8 transaction field index}
+- Pops: _None_
+- Pushes: any
+- push field from current transaction to stack
+
+Most fields are a simple copy of a uint64 or byte string value. `XferAsset` is the concatenation of the AssetID Creator Address (32 bytes) and the big-endian bytes of the uint64 AssetID Index for a total of 40 bytes.
+
+`txn` Fields:
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Sender | []byte |
+| 1 | Fee | uint64 |
+| 2 | FirstValid | uint64 |
+| 3 | LastValid | uint64 |
+| 4 | Note | []byte |
+| 5 | Receiver | []byte |
+| 6 | Amount | uint64 |
+| 7 | CloseRemainderTo | []byte |
+| 8 | VotePK | []byte |
+| 9 | SelectionPK | []byte |
+| 10 | VoteFirst | uint64 |
+| 11 | VoteLast | uint64 |
+| 12 | VoteKeyDilution | uint64 |
+| 13 | Type | []byte |
+| 14 | TypeEnum | uint64 |
+| 15 | XferAsset | []byte |
+| 16 | AssetAmount | uint64 |
+| 17 | AssetSender | []byte |
+| 18 | AssetReceiver | []byte |
+| 19 | AssetCloseTo | []byte |
+| 20 | GroupIndex | uint64 |
+
+
+TypeEnum mapping:
+
+| Index | Name |
+| --- | --- |
+| 0 | unknown |
+| 1 | pay |
+| 2 | keyreg |
+| 3 | acfg |
+| 4 | axfer |
+| 5 | afrz |
+
+
+## global
+
+- Opcode: 0x32 {uint8 global field index}
+- Pops: _None_
+- Pushes: any
+- push value from globals to stack
+
+`global` Fields:
+
+| Index | Name | Type |
+| --- | --- | --- |
+| 0 | Round | uint64 |
+| 1 | MinTxnFee | uint64 |
+| 2 | MinBalance | uint64 |
+| 3 | MaxTxnLife | uint64 |
+| 4 | TimeStamp | uint64 |
+| 5 | ZeroAddress | []byte |
+| 6 | GroupSize | uint64 |
+
+
+## gtxn
+
+- Opcode: 0x33 {uint8 transaction group index}{uint8 transaction field index}
+- Pops: _None_
+- Pushes: any
+- push field to the stack from a transaction in the current transaction group
+
+for notes on transaction fields available, see `txn`
+
+## load
+
+- Opcode: 0x34 {uint8 position in scratch space to load from}
+- Pops: _None_
+- Pushes: any
+- copy a value from scratch space to the stack
+
+## store
+
+- Opcode: 0x35 {uint8 position in scratch space to store to}
+- Pops: *... stack*, any
+- Pushes: _None_
+- pop a value from the stack and store to scratch space
+
+## bnz
+
+- Opcode: 0x40 {0..0x7fff forward branch offset, big endian}
+- Pops: *... stack*, uint64
+- Pushes: _None_
+- branch if value is not zero
+
+for a bnz instruction at `pc`, if the last element of the stack is not zero then branch to instruction at `pc + 3 + N`, else proceed to next instruction at `pc + 3`
```

Comment:
We reject if any jump would land not at the start of an instruction, right? That should probably be mentioned here.

Reply:
noted